conn_test.go

```diff
@@ -398,6 +452,252 @@ func testConnReadWatermarkFromBatch(t *testing.T, conn *Conn) {
 	batch.Close()
 }
 
+func createGroup(t *testing.T, conn *Conn, groupID string) (generationID int32, memberID string, stop func()) {
+	// join the group
+	joinGroup, err := conn.joinGroup(joinGroupRequestV2{
+		GroupID:          groupID,
+		SessionTimeout:   int32(time.Minute / time.Millisecond),
+		RebalanceTimeout: int32(time.Second / time.Millisecond),
+		ProtocolType:     "roundrobin",
+		GroupProtocols: []joinGroupRequestGroupProtocolV2{
+			{
+				ProtocolName:     "roundrobin",
+				ProtocolMetadata: []byte("blah"),
+			},
+		},
+	})
+	if err != nil {
+		t.Fatalf("bad joinGroup: %s", err)
+	}
+
+	// sync the group
+	_, err = conn.syncGroups(syncGroupRequestV1{
+		GroupID:      groupID,
+		GenerationID: joinGroup.GenerationID,
+		MemberID:     joinGroup.MemberID,
+		GroupAssignments: []syncGroupRequestGroupAssignmentV1{
+			{
+				MemberID:          joinGroup.MemberID,
+				MemberAssignments: []byte("blah"),
+			},
+		},
+	})
+	if err != nil {
+		t.Fatalf("bad syncGroups: %s", err)
+	}
+
+	generationID = joinGroup.GenerationID
+	memberID = joinGroup.MemberID
+	stop = func() {
+		conn.leaveGroup(leaveGroupRequestV1{
+			GroupID:  groupID,
+			MemberID: joinGroup.MemberID,
+		})
+	}
+
+	return
+}
+
+func testConnDescribeGroupRetrievesAllGroups(t *testing.T, conn *Conn) {
+	groupID := makeGroupID()
+	_, _, stop1 := createGroup(t, conn, groupID)
+	defer stop1()
+
+	out, err := conn.describeGroups(describeGroupsRequestV1{
+		GroupIDs: []string{groupID},
+	})
+	if err != nil {
+		t.Fatalf("bad describeGroups: %s", err)
+	}
+
+	if v := len(out.Groups); v != 1 {
+		t.Fatalf("expected 1 group, got %v", v)
+	}
+	if id := out.Groups[0].GroupID; id != groupID {
+		t.Fatalf("bad group: got %v, expected %v", id, groupID)
+	}
+}
+
+func testConnFindCoordinator(t *testing.T, conn *Conn) {
+	groupID := makeGroupID()
+
+	for attempt := 0; attempt < 10; attempt++ {
+		if attempt != 0 {
+			time.Sleep(time.Millisecond * 50)
+		}
+		response, err := conn.findCoordinator(findCoordinatorRequestV1{CoordinatorKey: groupID})
+		if err != nil {
+			switch err {
+			case GroupCoordinatorNotAvailable:
+				continue
+			default:
+				t.Fatalf("bad findCoordinator: %s", err)
+			}
+		}
+
+		if response.Coordinator.NodeID == 0 {
+			t.Fatalf("bad NodeID")
```

Comment:
Could you replace the use of Fatalf with Errorf? In general Fatal* is useful for errors that occur within the initialization phase of a test (because there's no way the test can succeed if a connection cannot be established for example).

However when verifying the state of the system after the execution of whatever code the test exercise, it's more interesting to report all that went wrong instead of just the first problem.

Reply:
Sure, happy to change to that.